### ISSUE TITLE
The user in the status bar is stuck at root

### ISSUE THREAD

**Issue by aofei** [gpakosz/.tmux#403]
## The problem

I just pulled the latest commit from the `master` branch. Then I found that if I execute any `sudo`-prefixed command (such as `sudo whoami`), the user in the status bar will remain `root!`, even if I've returned to a standard user.

## My setup

```bash
$ uname -a
Linux as-server 5.8.7-arch1-1 #1 SMP PREEMPT Sat, 05 Sep 2020 12:31:32 +0000 x86_64 GNU/Linux

$ tmux -V
tmux 3.1b

$ ssh -V
OpenSSH_8.3p1, OpenSSL 1.1.1g  21 Apr 2020
```

**Comment by gpakosz** [comment]
Hello @aofei 👋

I can't reproduce the problem here:
```
$ uname -a
Linux mustafar 5.8.9-arch2-1 #1 SMP PREEMPT Sun, 13 Sep 2020 23:44:55 +0000 x86_64 GNU/Linux
```
```
$ tmux -V
tmux 3.1b
``` 

After having pulled `master`, did you reload the configuration?

Before reloading the configuration can you please do a thing or two?
- Can you please paste the output of
  ```
  $ tmux show -gv status-right
  ```
- In a pane where you're **not** root but the status bar still displays `root!`, can you please do
  ```
  $ cut -c3- ~/.tmux.conf | sh -sx _username
  ```

**Comment by aofei** [comment]
Yes, I reloaded the configuration (`C-a` + `r`) and the problem exists.

However, I just killed the tmux session and recreated a new one, then the problem disappeared. So it's not a bug.

Sorry for wasting your time. Next time I'll also remember to restart the session before giving a feedback. 😥

**Comment by gpakosz** [comment]
You shouldn't have to restart the session.

I'll investigate more on this issue.

**Comment by aofei** [comment]
Okay, I'll pay attention. If I encounter it again, I'll report back here.

**Comment by aofei** [comment]
Yes it happened again.

```bash
$ tmux show -gv status-right
#[fg=#080808,bg=#080808,none]#[fg=#8a8a8a,bg=#080808,none]#[fg=none]#[bg=none]#[none]#{?client_prefix,⌨ ,}#[fg=none]#[bg=none]#[none]#{?session_many_attached,👓 ,}#[fg=none]#[bg=none]#[none]#{?pane_synchronized,🔒,} #{?battery_status, #{battery_status},}#{?battery_bar, #{battery_bar},}#{?battery_percentage, #{battery_percentage},} #[fg=#8a8a8a,bg=#080808,none] %R #[fg=#8a8a8a,bg=#080808,none] %d %b #[fg=#d70000,bg=#080808,none]#[fg=#e4e4e4,bg=#d70000,none] #(cut -c3- ~/.tmux.conf | sh -s _username #{pane_tty} false #D)#[fg=none]#[bg=none]#[bold,blink]#{?#{==:#(cut -c3- ~/.tmux.conf | sh -s _username #{pane_tty} #D),root},!,}#[default]#[fg=#e4e4e4,bg=#d70000,none] #[fg=#e4e4e4,bg=#d70000,none]#[fg=#000000,bg=#e4e4e4,bold] #(cut -c3- ~/.tmux.conf | sh -s _hostname #{pane_tty} false #h #D)
```

and

```bash
$ cut -c3- ~/.tmux.conf | sh -sx _username                                                                                                                                
+ :
+ set -e
+ unset GREP_OPTIONS
+ export LC_NUMERIC=C
+ LC_NUMERIC=C
+ printf ''
+ sed -E s///
++ uname -s
+ _uname_s=Linux
+ command -v pkill
+ _username
++ tmux display -p '#{pane_tty}'
+ tty=/dev/pts/0
+ ssh_only=
++ _tty_info /dev/pts/0
++ tty=pts/0
++ case "$_uname_s" in
++ ps -t pts/0 -o user=XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX -o pid= -o ppid= -o command=
++ awk '
        NR > 1 && ((/ssh/ && !/-W/) || !/ssh/) {
          user[$2] = $1; parent[$2] = $3; child[$3] = $2; pid=$2; $1 = $2 = $3 = ""; command[pid] = substr($0,4)
        }
        END {
          for (i in parent)
          {
            j = i
            while (parent[j])
              j = parent[j]

            if (!(i in child) && j != 1)
            {
              print i":"user[i]":"command[i]
              exit
            }
          }
        }
      '
+ tty_info='584228:aofei:ps -t pts/0 -o user=XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX -o pid= -o ppid= -o command='
+ command='aofei:ps -t pts/0 -o user=XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX -o pid= -o ppid= -o command='
+ command='ps -t pts/0 -o user=XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX -o pid= -o ppid= -o command='
++ _ssh_or_mosh_args 'ps -t pts/0 -o user=XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX -o pid= -o ppid= -o command='
++ case "$1" in
++ printf %s ''
+ ssh_or_mosh_args=
+ '[' -n '' ']'
+ _is_enabled ''
+ '[' x = xtrue ']'
+ '[' x = xyes ']'
+ '[' x = xenabled ']'
+ '[' x = x1 ']'
+ username='aofei:ps -t pts/0 -o user=XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX -o pid= -o ppid= -o command='
+ username=aofei
+ printf '%s\n' aofei
aofei
```

and the screenshot

![C3C195FE-A256-4846-9934-121CC769BF0E](https://user-images.githubusercontent.com/5037285/93725054-a94e1680-fbde-11ea-8f36-43319a2b97c6.png)

**Comment by gpakosz** [comment]
When it happens, does `tmux refresh-client -S` help?

**Comment by aofei** [comment]
`tmux refresh-client -S` helped.

Sorry I haven't caught how it's triggered yet. But once it's triggered, executing any `sudo`-prefixed command in that session will cause problems. Then I have to restart the session or execute `tmux refresh-client -S`.

**Comment by gpakosz** [comment]
Well there's necessarily a lag between when you're no more root and when tmux decides to refresh the status line

What's the value of your `tmux_conf_theme_status_left` and `tmux_conf_theme_status_right` variables?

**Comment by gpakosz** [comment]
Also, tmux I see you have the prefix indicator in `tmux_conf_theme_status_right` which means tmux should refresh the status line when you hit the prefix key

**Comment by aofei** [comment]
Both `tmux_conf_theme_status_left` and `tmux_conf_theme_status_right` are the default values ​​specified by this project.

In fact, it's not a lag. It'll always be stuck at `root!` in that panel, no matter what command I execute later (or switch between different windows or panels). Unless I close that panel and create a new one. But if I execute a `sudo`-prefixed command in the new panel, the problem will show up again.

**Comment by gpakosz** [comment]
Ok I will keep investigating

**Comment by gpakosz** [comment]
To further investigate, when it happens can you please

1. Describe the sequence of commands you launched to make it stuck, ideally a minimal repro

2. Get the current pane PID (the pane for which the status bar doesn't update)
  ```
  $ tmux display -p '#{pane_pid}'
  ```

3. Use `ps -t` with the value returned by the previous command, e.g.
  ```
  $ ps -t pts/0 -o user=XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX -o pid= -o ppid= -o command=
  ```

And paste the output?

**Comment by aofei** [comment]
Ok, I'll pay attention next time it happens.

**Comment by gpakosz** [comment]
BTW what's the output of
```
$ tmux show -gv status-inverval
```

**Comment by aofei** [comment]
```bash
$ tmux show -gv status-interval
10
```

**Comment by gpakosz** [comment]
Hey @aofei  👋

Any news?

**Comment by aofei** [comment]
It happened several times recently. Unfortunately, I still don't know what specific steps are causing it. It's really cunning... 😥

But anyway, I'll still pay attention to it.

**Comment by LubosRemplik** [comment]
I can confirm having same problem, only with first pane once I open more I am seeing right users

Anything I can do to help reproduce this?

**Comment by aofei** [comment]
I still encounter this from time to time. But sadly, I still cannot find a way to reproduce it. This's a tricky one.

**Comment by gpakosz** [comment]
Hello @LubosRemplik 👋

Which OS are you using? And is it happening with the current unmodified `master` branch?

**Comment by LubosRemplik** [comment]
- Ubuntu 20.04
- Yep, using `master` branch, cloned today (last commit 7a71a82)

I did some modification in .tmux.config.local, but nothing big so pretty much what I can from clone + install instrucntions

I have just find out when it happens, not sure if @aofei is using that too, but I am using [`mosh`](https://mosh.org/) instead of `ssh` and it happens only when i do `mosh lubos@ubuntu2004`, when i do `ssh lubos@ubuntu2004` it seems to be fine = maybe there is no problem on tmux side, but it is mosh

**Comment by gpakosz** [comment]
@LubosRemplik but when you do `lubos@ubuntu2004`, is the reported user `root`?

**Comment by LubosRemplik** [comment]
> @LubosRemplik but when you do `lubos@ubuntu2004`, is the reported user `root`?

Yes, when I do `tmux` reported user is `root` in status line

**Comment by gpakosz** [comment]
That's weird. Without using `sudo`?

**Comment by LubosRemplik** [comment]
> That's weird. Without using sudo? Also, do you have a stable repro?

No `sudo` used

More findings 

- `mosh lubos@ubuntu2004 tmux`, reported user `root`
- `mosh --no-ssh-pty lubos@ubuntu2004 tmux`, reported user `lubos`
- `ssh -t lubos@ubuntu2004 tmux`, reported user `lubos`
- `ssh lubos@ubuntu2004 tmux`, i am getting `open terminal failed: not a terminal`

I am running `ssh client` and `mosh client` on [WSL2 Ubuntu](https://docs.microsoft.com/en-us/windows/wsl/install-win10) and login to the remote VPS ubuntu2004 with mosh-server and ssh daemon. Also I am using [windows terminal](https://github.com/microsoft/terminal) 

Now I think I should open issue at [mosh github](https://github.com/mobile-shell/mosh/issues?q=is%3Aissue+is%3Aopen+root+tmux)

**Comment by gpakosz** [comment]
> Now I think I should open issue at mosh github

Not yet

**Comment by gpakosz** [comment]
So you have a repro.

Can you please in a pane do
```
$ tmux display -p '#{pane_tty}'
$ mosh lubos@ubuntu2004 tmux
```
The first command should print something like `/dev/pts/XX`
In another pane, do
```
ps -t /dev/pts/XX -o user=XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX -o pid= -o ppid= -o command=
```

**Comment by LubosRemplik** [comment]
I had to do another way around

```
$ mosh lubos@ubuntu2004 tmux
$ tmux display -p '#{pane_tty}'
```

since tmux is at the remote ubuntu2004 server  
and getting back  
`/dev/pts/0`

splitted the widnow (= new pane if i am not wrong) and did
```
ps -t /dev/pts/XX -o user=XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX -o pid= -o ppid= -o command=
```
(not sure if I should replace X somehow) and return is

```
error: TTY could not be found

Usage:
 ps [options]

 Try 'ps --help <simple|list|output|threads|misc|all>'
  or 'ps --help <s|l|o|t|m|a>'
 for additional help text.

For more details see ps(1).
```

**Comment by LubosRemplik** [comment]
When i do with `mosh --no-ssh-pty lubos@ubuntu2004 tmux` I am getting same results (well it returns `/dev/pts/1`) but second result is same

**Comment by gpakosz** [comment]
You need to get the pane tty before launching `mosh`

**Comment by gpakosz** [comment]
Wait so you have no local tmux, but only tmux launched on the remote server?

**Comment by gpakosz** [comment]
You should replace `/dev/pts/XX` by `/dev/pts/0` 🙂

**Comment by LubosRemplik** [comment]
> Wait so you have no local tmux, but only tmux launched on the remote server?

Exactly

Output of above is

```
XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
root                               9113    8366 node
root                               9207    8390 node
lubos                           3498204 3497953 -bash
lubos                           3498924 3498204 ps -t /dev/pts/0 -o user=XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX -o pid= -o ppid= -o command=
```

**Comment by LubosRemplik** [comment]
Seems you were expecting this output `lubos`, but still `root!` in status bar, see screenshot

![image](https://user-images.githubusercontent.com/556169/108471742-21069700-728c-11eb-8d5c-1b1a05723daa.png)

**Comment by LubosRemplik** [comment]
Second pane with right user, while when I switch to the left pane `root!` is there

![image](https://user-images.githubusercontent.com/556169/108472051-85c1f180-728c-11eb-9fc7-5819a06b7185.png)

**Comment by gpakosz** [comment]
Can you compare the outputs of `ps -t` in each pane, I'm curious about whether a `node` root process appears in each output.

These process owned by `root` is the cause for the bug.
Can you please paste the full output of `ps -t` in a text file, so that I can fake a `ps` command and troubleshoot on my side?

**Comment by LubosRemplik** [comment]
You are right,

left pane shows `root ... node` output  
`ps -t /dev/pts/0 -o user=XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX -o pid= -o ppid= -o command= > left_pane.txt`  
[left_pane.txt](https://github.com/gpakosz/.tmux/files/6008494/left_pane.txt)

right pane shows NO `root ... node` output    
`ps -t /dev/pts/2 -o user=XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX -o pid= -o ppid= -o command= > right_pane.txt`  
[right_pane.txt](https://github.com/gpakosz/.tmux/files/6008493/right_pane.txt)

I can confirm there are some processess with node under root (obviously), but how can get rid of these with above command which obvioulsly shows the username for the status bar?

**Comment by gpakosz** [comment]
Well on my side I can start thinking about how to improve the detection logic.
But the initial culprit seems to be there's a `node` process attached to this TTY 🤷‍♂️.

What node based apps are you using?

**Comment by LubosRemplik** [comment]
Well I am running all nodes (i think) via docker container, see ps output

[ps_grep_node.txt](https://github.com/gpakosz/.tmux/files/6008563/ps_grep_node.txt)

**Comment by LubosRemplik** [comment]
Using `node` to build assets (css, js) which is unrelated to this, but explains a bit why `node` is in `ps -t` output

**Comment by gpakosz** [comment]
That doesn't explain 2 things though
- why it's owned by `root`
- why it's (supposedly) stalled?

**Comment by LubosRemplik** [comment]
When i kill those 2 node processess, they are comming back.

Wondering if tmux is reading this `ps -t` output and using in status bar, maybe better if it reads `command` too, meaning when command == tmux, the user == lubos

**Comment by gpakosz** [comment]
> When i kill those 2 node processess, they are comming back.

What's your terminal emulator?
Can you use `pstree` to see what's the parent process of these `node` process?

**Comment by LubosRemplik** [comment]
> > When i kill those 2 node processess, they are comming back.
> 
> What's your terminal emulator?
> Can you use `pstree` to see what's the parent process of these `node` process?

I use [Windows terminal](https://github.com/microsoft/terminal), the pstree output attached  
[pstree.txt](https://github.com/gpakosz/.tmux/files/6008640/pstree.txt)

**Comment by LubosRemplik** [comment]
> Wondering when tmux is reading this ps -t output and using in status bar, maybe better if it reads command too, meaning when command == tmux, the user == lubos

https://github.com/gpakosz/.tmux/blob/master/.tmux.conf#L524

Would it be bad idea to replace that line with 

ps -t "$tty" -o user=XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX -o pid= -o ppid= -o command= | grep tmux | awk '

Worth PR?

**Comment by LubosRemplik** [comment]
Well above does not work, but let me know if I should try to adjust snippet part from line 524 since I am having this "wrong" ps -t output

**Comment by gpakosz** [comment]
> Would it be bad idea to replace that line with
> ```ps -t "$tty" -o user=XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX -o pid= -o ppid= -o command= | grep tmux | awk '```
> Worth PR?

That's not the way to fix it

**Comment by gpakosz** [comment]
Ideally what you can do is finding out the reason for these `node` processes to be there

**Comment by gpakosz** [comment]
@LubosRemplik can you please give me the output of
```
$ ps -t <tty> -o pid= -o ppid= -o tty= -o cmd= -o s=
```
where `<tty>` corresponds to the pane for which there are `node` processes owned by `root`?

**Comment by LubosRemplik** [comment]
> -o pid= -o ppid= -o tty= -o cmd= -o s=

output is 

```
3723578 3723561 pts/0    node                        S
3724927 3724905 pts/0    node                        S
4106494 4106248 pts/0    -bash                       S
4108752 4106494 pts/0    ps -t /dev/pts/0 -o pid= -o R
```

**Comment by gpakosz** [comment]
Can you have a look at what's PID `3723561`?
Maybe with
```
$ ps aux | grep 3723561
```

**Comment by LubosRemplik** [comment]
> Can you have a look at what's PID `3723561`?
> Maybe with
> 
> ```
> $ ps aux | grep 3723561
> ```

```
root     3723561  0.0  0.0 108720  5336 ?        Sl   09:28   0:00 containerd-shim -namespace moby -workdir /var/lib/containerd/io.containerd.runtime.v1.linux/moby/0b20d4c75d9695ed0af6a0537b777fa47244cd2838be1f673807f648d0bb0e3f -address /run/containerd/containerd.sock -containerd-binary /usr/bin/containerd -runtime-root /var/run/docker/runtime-runc
```

**Comment by gpakosz** [comment]
Can it be the default command launched by Docker, as specified by the `Dockerfile`?

**Comment by LubosRemplik** [comment]
Well not sure what it is, what it was, but now I have stopped and removed all docker containers and this node process is gone. So it was caused by some docker container with node, but cant reproduce now since it is gone.

Later on when I start some container and see this info, I might tell you more info here

**Comment by gpakosz** [comment]
In fact when you reproduce it again, I'm interested in
```
$ ps -t <tty> -o user= -o pid= -o ppid= -o tty= -o cmd= -o stat=
```
With that I may be able to filter out background processes

**Comment by gpakosz** [comment]
I pushed the `gh-403` branch which may solve this issue.

Please keep me posted!

**Comment by LubosRemplik** [comment]
Thanks! I'll try the branch when the problem occur again and will let you know

**Comment by LubosRemplik** [comment]
Unfortunatelly fix at `gh-403` does not work yet.
I have tested on another machine where I installed tmux3.0a, used your config and this machine is running docker containers including node too. See below output

```
lubos@debra:~$ ps -t /dev/pts/0 -o user= -o pid= -o ppid= -o tty= -o cmd= -o stat=                                                                                                                                                              
root        3371    2829 pts/0    node                        Ssl+
lubos     770071  769800 pts/0    -bash                       Ss
lubos     781952  770071 pts/0    ps -t /dev/pts/0 -o user= - R+
lubos@debra:~$ cd .tmux
lubos@debra:~/.tmux$ git checkout gh-403
Branch 'gh-403' set up to track remote branch 'gh-403' from 'origin'.
Switched to a new branch 'gh-403'
lubos@debra:~/.tmux$ git pull origin gh-403
From https://github.com/gpakosz/.tmux
 * branch            gh-403     -> FETCH_HEAD
Already up to date.
lubos@debra:~/.tmux$ cd ..
lubos@debra:~$ ls -la .tmux.conf
lrwxrwxrwx 1 lubos admin 28 Feb 23 16:13 .tmux.conf -> /home/lubos/.tmux/.tmux.conf
lubos@debra:~$ tmux source-file .tmux.conf
```

screenshot

![image](https://user-images.githubusercontent.com/556169/108978350-c903d200-7689-11eb-9abd-45fa59b80d95.png)

**Comment by gpakosz** [comment]
What I still don't get is why the `node` process uses the same TTY as the `bash` process.

Can you please detail in which order you launch things?

**Comment by LubosRemplik** [comment]
I did

`mosh lubos@debra -- tmux new-session -As lubos`

then i saw `root` user in status-right (sorry custom .tmux.config.local using `#{username}@#H`) so I did following to report

```
lubos@debra:~$ ps -t /dev/pts/0 -o user= -o pid= -o ppid= -o tty= -o cmd= -o stat=                                                                                                                                                              
root        3371    2829 pts/0    node                        Ssl+
lubos     770071  769800 pts/0    -bash                       Ss
lubos     781952  770071 pts/0    ps -t /dev/pts/0 -o user= - R+
lubos@debra:~$ cd .tmux
lubos@debra:~/.tmux$ git checkout gh-403
Branch 'gh-403' set up to track remote branch 'gh-403' from 'origin'.
Switched to a new branch 'gh-403'
lubos@debra:~/.tmux$ git pull origin gh-403
From https://github.com/gpakosz/.tmux
 * branch            gh-403     -> FETCH_HEAD
Already up to date.
lubos@debra:~/.tmux$ cd ..
lubos@debra:~$ ls -la .tmux.conf
lrwxrwxrwx 1 lubos admin 28 Feb 23 16:13 .tmux.conf -> /home/lubos/.tmux/.tmux.conf
lubos@debra:~$ tmux source-file .tmux.conf
```

**Comment by gpakosz** [comment]
And `tmux` runs inside a Docker container? Does `mosh` have a dependency on `node`?

**Comment by LubosRemplik** [comment]
Nope, `tmux` does not run in container, it is installed in Ubuntu 20,04 via apt.

From https://mosh.org/ dependencies section, `node` is not listed there.

I am sure if I kill running docker containers (the once with node image), it will disappear again. Below is list of node containers running at debra server

```
6f3985cad0b3        37f455de4837                       "sh -c './node_modul…"   4 weeks ago         Up 7 hours                      0.0.0.0:35120->35729/tcp                         thomasmathys_gulp_1
6473b935adb1        37f455de4837                       "sh -c './node_modul…"   4 weeks ago         Up 7 hours                                                                       thomasmathys_webpack-admin_1
3f602bc22dd0        b328632eb00c                       "docker-entrypoint.s…"   2 months ago        Up 7 hours                                                                       pf_node_1
faab570d05c2        37f455de4837                       "node"                   3 months ago        Up 7 hours                                                                       shome_node_1
5358fb3734fb        node:12-alpine                     "docker-entrypoint.s…"   4 months ago        Up 7 hours                                                                       amexpp_gulp_1
```

When I kill 4th one, `node` disappear from ps, when I start docker container with node again, it is in ps but 2nd item so not in `#{username}`

```
lubos     770071  769800 pts/0    -bash                       Ss
root     1178039 1178002 pts/0    node                        Ssl+
lubos    1179412  770071 pts/0    ps -t /dev/pts/0 -o user= - R+
```

When i look to docker-compose yaml, which runs docker node container, i see the problem (screenshot)

![image](https://user-images.githubusercontent.com/556169/108983777-7fb68100-768f-11eb-9375-4a84cf8b5140.png)

So related to docker + tty config if you see what I mean.

**Comment by LubosRemplik** [comment]
Silly question, why not to use `whoami` isntead parsing `ps -t /dev/pts/0 -o user= -o pid= -o ppid= -o tty= -o cmd= -o stat= ` output?

**Comment by gpakosz** [comment]
> Silly question, why not to use `whoami` instead parsing `ps -t /dev/pts/0 -o user= -o pid= -o ppid= -o tty= -o cmd= -o stat=` output?

Because I want the `user` of the running command. So that it displays e.g. `root` for commands running with `sudo`.
The problem in your case is that there are different process, running from different users, using the same tty at the same time

**Comment by gpakosz** [comment]
Did you run `docker-compose` from the `bash` session you're attaching to?

**Comment by LubosRemplik** [comment]
> Did you run docker-compose from the bash session you're attaching to?

First time I saw this problem I did not = it was there when I logged into the server

When did tests above, executed `docker-compose` from bash session

**Comment by gpakosz** [comment]
So far I don't understand why it's the same TTY

**Comment by LubosRemplik** [comment]
From reading online, `docker -t` or docker-compose tty option is Pseudo-TTY.

> use -t, usually in the combination -it and you have a terminal driver added, It basically makes the container start look like a terminal connection session.

Not sure if it explains it?

> > Silly question, why not to use `whoami` instead parsing `ps -t /dev/pts/0 -o user= -o pid= -o ppid= -o tty= -o cmd= -o stat=` output?
> 
> Because I want the `user` of the running command. So that it displays e.g. `root` for commands running with `sudo`.
> The problem in your case is that there are different process, running from different users, using the same tty at the same time

OK. not sure if I follow, but `whoami` returns `lubos` and 'sudo whoami` returns `root`. Are you saying user would not change when running sudo command?

**Comment by gpakosz** [comment]
> OK. not sure if I follow, but `whoami` returns `lubos` and `sudo whoami` returns `root`. Are you saying user would not change when running sudo command?

How do you exactly run `whoami` for an existing tmux pane?

**Comment by LubosRemplik** [comment]
> > OK. not sure if I follow, but `whoami` returns `lubos` and `sudo whoami` returns `root`. Are you saying user would not change when running sudo command?
> 
> How do you exactly run `whoami` for an existing tmux pane?

Well since new to tmux, this is beyond my knowledge yet. Just ignore me :)

**Comment by gpakosz** [comment]
But are you interacting with this `node` process? Or is it just there in the background? Why do you need to allocate a pseudo TTY in the first place?

**Comment by LubosRemplik** [comment]
> But are you interacting with this `node` process? Or is it just there in the background? Why do you need to allocate a pseudo TTY in the first place?

It is just in background and I do not need allocate the pseudo TTY (removed that config from docker-compose.yml), but still good to know this might be the problem...

Maybe just note to readme about this is enough for now

**Comment by gpakosz** [comment]
Can you reproduce it without `docker-compose`but with `docker run -t` instead?

**Comment by LubosRemplik** [comment]
Yes I can reproduce with `docker run -it node`, when running I can see `node` process in `ps -t /dev/pts/0 -o user= -o pid= -o ppid= -o tty= -o cmd= -o stat=  ` list.  When stopping, it is gone.

Note: it is third row so it does not cause problem with `#{username}` but once it gets to the top, it is there.

My output:
```
lubos@debra:~/vim/src$ ps -t /dev/pts/0 -o user= -o pid= -o ppid= -o tty= -o cmd= -o stat=
lubos     770071  769800 pts/0    -bash                       Ss
lubos    1546804  770071 pts/0    ssh oiserver                S+
root     1622663 1622640 pts/0    node                        Ssl+
lubos@debra:~/vim/src$ ps -t /dev/pts/0 -o user= -o pid= -o ppid= -o tty= -o cmd= -o stat=
lubos     770071  769800 pts/0    -bash                       Ss
lubos    1546804  770071 pts/0    ssh oiserver                S+
lubos@debra:~/vim/src$
```

**Comment by gpakosz** [comment]
Now I'm confused about the `ssh` process here and why is mapped to `pts/0`

**Comment by gpakosz** [comment]
When you create different tmux panes and in each you do `$ tmux display -p '#{pane_tty}'`, what do the different panes say?

**Comment by LubosRemplik** [comment]
> Now I'm confused about the ssh process here and why is mapped to pts/0

sorry about that, you can ignore that - it is from another window when I did ssh somewhere else. I adjusting above as when I exit that ssh connection, it is gone

> $ tmux display -p '#{pane_tty}'

one pane `/dev/pts/0` another one `/dev/pts/2`

**Comment by gpakosz** [comment]
Hey @aofei and @LubosRemplik 👋

I just force pushed the `gh-403` branch. Can you please give it a try? Please make sure you reload the configuration after switching branch.
